Added Nil Check for Package in File Descriptor

tmulkern opened the pull request:
## What?

This code is to guard against a `runtime error: invalid memory address or nil pointer dereference` occuring

## Why?

The package specifier is optional in protobuf spec https://protobuf.dev/programming-guides/proto3/#packages

## Checklist

<!-- 
If you haven't read the contributing guidelines https://github.com/grafana/k6/blob/master/CONTRIBUTING.md 
and code of conduct https://github.com/grafana/k6/blob/master/CODE_OF_CONDUCT.md yet, please do so
-->

- [x] I have performed a self-review of my code.
- [ ] I have added tests for my changes. N/A
- [x] I have run linter locally (`make ci-like-lint`) and all checks pass. N/A
- [x] I have run tests locally (`make tests`) and all tests pass. N/A
- [ ] I have commented on my code, particularly in hard-to-understand areas. N/A
<!-- - [ ] Any other relevant item -->

## Related PR(s)/Issue(s)

<!-- - <https://github.com/grafana/...> -->

<!-- Does it close an issue? -->
Fixes issue #3222 
<!-- Closes #ISSUE-ID -->

<!-- Thanks for your contribution! 🙏🏼 -->

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/grafana/k6?pullRequest=3223) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/grafana/k6?pullRequest=3223) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/grafana/k6?pullRequest=3223) it.</sub>

mstoykov reviewed (approved):
LGTM!

Sorry for the slow review :bow: . 

Due to vacations across the team and the automatic assigning not taking this into account, sometimes things get assigned to people who are on vacation :(.

Unfortunately we are also at the end of the development cycle and have some higher priorities things we want to make certain do work for v0.46.0.

The fix seems pretty safe and I can't see any problems with it, but I still prefer if @olegbespalov gives it a :+1: specifically. This might mean that we do merge this in the codefreeze, but given the size and the fact it is a bug that doesn't seem like that much of a problem.

@tmulkern Can you please add a small test that panics without this change though.

Also we will need forward/backport this to [xk6-grpc](https://github.com/grafana/xk6-grpc/). 

I will move this back to v0.46.0 milestone for now, but I can't promise we will actually make it :(

Thanks for the bug report and fix :bow: And once again sorry for the slow reply

tmulkern commented:
@mstoykov no problem on the response, I am in the same situation myself in my day job, it is the time for Vacations!!

tmulkern commented:
Regarding the test, no problem in creating, but do we have a GRPC Test Service/Mock Service that I can reference. I couldn't find any. I am not *that* familiar with the code base, so I could be blind to it ;-)

codebien reviewed (changes requested):
Hey @tmulkern,
thanks for your contribution. :bow: 

>Regarding the test, no problem in creating, but do we have a GRPC Test Service/Mock Service that I can reference

I guess it should be enough to add a new test case to this test table. I think we should cover also the case with clashing names now that we are skipping the package.

https://github.com/grafana/k6/blob/3c006a5e7b4a44adbc85aefc1933759d22b99286/lib/netext/grpcext/conn_test.go#L160

tmulkern commented:
Hey @codebien 

Added tests for following scenarios

* Services with no package - "NoPackage"
* Services with no package and duplicated - "NoPackageDeduplicateServices"
* One services with a package, one without - "MixPackage"
* Two services with a package (Duplicate), one without - "MixPackageDeduplicateServices"

Also updated the TestResolveFileDescriptors to handle cases with no packages

olegbespalov commented on lib/netext/grpcext/conn_test.go:
```diff
@@ -199,10 +221,16 @@ func TestResolveFileDescriptors(t *testing.T) {
+			}
+
 			for i, service := range tt.services {
 				// if only one package is defined then
 				// the package is the same for every service
```
```suggestion
```

olegbespalov reviewed (approved):
To be honest, in that particular case, I'm more up for keeping the logic (with the comment), like

```golang
pkg := ""

// if only one package is defined then
// the package is the same for every service
if len(tt.pkgs) == 1 {
   pkg = tt.pkgs[0]
}
```

The rationale is that having the empty strings confuses me more than the logic of the comment.

tmulkern commented:
Applogies people for the late reply, as mstoykov alluded to it is vacation season and I was away on vacation

olegbespalov commented:
Hi @tmulkern !

I am sharing this here for the sake of transparency. We have another PR related to the [support of gRPC's reflection v1](https://github.com/grafana/k6/pull/3338). 

I believe that PR also fixes the issue that you reported. So we will probably close this PR in favor of https://github.com/grafana/k6/pull/3338.

Anyway, thank you for reporting and contribution! :bow:

tmulkern commented:
Hey [olegbespalov](https://github.com/olegbespalov), thank's for the update. Sorry, life got in the way the 2 months and I hadn't got round to finishing this PR.
